0012/README.md

```diff
@@ -0,0 +1,215 @@
+
+We have added workflow notifications as an example:
+- [workflow-notifications.sql](workflow-notifications.sql): SQL script to setup notifications for workflow table
+- [listener.go](listener.go) - a listener watching for the notifications
+- [event.log](event.log) - log for INSERT, UPDATE and DELETE notifications
+
+## The API
+
+### Data Model
+
+ResourceType - a resource that an event can be associated with
+
+- Template
+- Hardware
+- Workflow
+
+EventType - an event type in tinkerbell space(a non-exhaustive list)
+
+- CREATED
+- UPDATED
+- DELETED
+- WORKFLOW_STARTED
+- WORKFLOW_INPROGRESS
+- WORKFLOW_FAILED
+- WORKFLOW_TIMEOUT
+
+Event - an event in tinkerbell space; and has the following structure:
+
+```
+{
+  "id": "uuid",
+  "resourceID": "uuid",
+  "resourceType": "workflow",
+  "eventType": "created",
+  "time": "",
+  "data": {
+  }
+}
+```
+
+- `id`: a unique identifier (UUID) for the event occured in tinkerbell space
+- `resourceID`: a unique identifier (UUID) for the resource, the event is associated with.
+  For example, `resourceID` will be set to `workflowID` for workflow resource and to `templateID` for the template resource.
+- `resouceType`: type of the resource (template, hardware, workflow) for which the event was generated
+- `eventType`: the event verb; describing the action on resource that generated the event
+- `time`: the event timestamp
+- `data`: the primary event payload, represented as `interface{}`.
+  For a workflow created event, the payload (data) can be the complete workflow structure.
+  For an on-timeout event the data can be an action that needs to be executed next.
+
+### EventClient
+
+A new `EventClient` should be developed with the primitive required for the events at least `Watch`.
+At the beginning, only a `Watch` function is required but we think a natural evolution will be to serve a function that can be used from other components to fire events.
+Or maybe we can do a `Watch` and `Create` straight away.
+All the resource clients will have a `Watch` function that will stream events for that particular resource.
+
+```
+Watch(ctx context.Context, in EventWatchRequest)
+```
+
+Where the `EventWatchRequest` can take the following structure:
+
+```
+type EventWatchRequest {
+    EventName  // the name of the event to filter by
+    ResourceID // the resource ID (ResourceType required when this is set)
+    ResourceType // workflow, template, action, hardware
+}
+```
+
+So when we create a `HardwareClient`, for instance, it would return a watcher for the `Hardware` resource type.
+All the clients like `HardwareClient` will be using `EventClient` under the hood.
+
+There is no logic to keep track of which events are fired or sent to a consumer.
+Every consumer when it connects to a stream of events will specify how old the events returned should be (by default 5m).
```

Comment:
The stream isn't being persisted in any of the examples here. That should be addressed in this proposal no?

Reply:
I have updated the proposal with details about the events table, where all the generated events will be stored.